RedBeanPHP/OODBBean.php

```diff
@@ -1066,126 +1066,96 @@ public function isListInExclusiveMode( $listName )
+				$this->__info['tainted'] = TRUE;
+				if ( isset( $this->__info["sys.parentcache.$property"] ) ) {
+					$bean = $this->__info["sys.parentcache.$property"];
 				} else {
-					$type = $property;
-				}
-				$bean = NULL;
-				if ( !is_null( $this->properties[$fieldLink] ) ) {
-					$bean = $redbean->load( $type, $this->properties[$fieldLink] );
-					//If the IDs dont match, we failed to load, so try autoresolv in that case...
-					if ( $bean->id !== $this->properties[$fieldLink] && self::$autoResolve ) {
-						$type = $this->beanHelper->getToolbox()->getWriter()->inferFetchType( $this->__info['type'], $property );
-						if ( !is_null( $type) ) {
-							$bean = $redbean->load( $type, $this->properties[$fieldLink] );
-							$this->__info["sys.autoresolved.{$property}"] = $type;
+					if ( isset( self::$aliases[$property] ) ) {
+						$type = self::$aliases[$property];
+					} elseif ( $this->fetchType ) {
+						$type = $this->fetchType;
+						$this->fetchType = NULL;
+					} else {
+						$type = $property;
+					}
+					$bean = NULL;
+					if ( !is_null( $this->properties[$fieldLink] ) ) {
+						$bean = $redbean->load( $type, $this->properties[$fieldLink] );
+						//If the IDs dont match, we failed to load, so try autoresolv in that case...
+						if ( $bean->id !== $this->properties[$fieldLink] && self::$autoResolve ) {
+							$type = $this->beanHelper->getToolbox()->getWriter()->inferFetchType( $this->__info['type'], $property );
+							if ( !is_null( $type) ) {
+								$bean = $redbean->load( $type, $this->properties[$fieldLink] );
+								$this->__info["sys.autoresolved.{$property}"] = $type;
+							}
 						}
 					}
 				}
+				$this->properties[$property] = $bean;
+				$returnValue = &$this->properties[$property];
+				break;
 			}
-			$this->properties[$property] = $bean;
-			$this->withSql               = '';
-			$this->withParams            = array();
-			$this->aliasName             = NULL;
-			$this->fetchType             = NULL;
-			$this->noLoad                = FALSE;
-			$this->all                   = FALSE;
-			$this->via                   = NULL;
-			return $this->properties[$property];
 
-		}
-		//Implicit: elseif ( $isOwn || $isShared ) {
-		if ( $this->noLoad ) {
-			$beans = array();
-		} elseif ( $isOwn ) {
-			$beans = $this->getOwnList( $listName, $redbean );
-		} else {
-			$beans = $this->getSharedList( lcfirst( substr( $property, 6 ) ), $redbean, $toolbox );
-		}
+			//Implicit: elseif ( $isOwn || $isShared ) {
+			if ( $this->noLoad ) {
+				$beans = array();
+			} elseif ( $isOwn ) {
+				$beans = $this->getOwnList( $listName, $redbean );
+			} else {
+				$beans = $this->getSharedList( lcfirst( substr( $property, 6 ) ), $redbean, $toolbox );
+			}
+
+			$this->properties[$property]          = $beans;
+			$this->__info["sys.shadow.$property"] = $beans;
+			$this->__info['tainted']              = TRUE;
+			$returnValue = &$this->properties[$property];
+
+		} while ( FALSE );
```

Comment:
Why we have a `do {} while ()` if you want to do it only once?

Reply:
The point was to reduce code and make it more readable by reseting modifiers and returning in one place only. The `do {} while ()` was a trick to act like a goto, doing the whole thing only once if it went to the end but brealing out in the middle if necessary.

Anyway I added a commit modifying things (and getting rid of the do-while) which I think is better.